Add Constraint Builder

s9latimm opened the pull request:
- [x] Implementation
  - [x] Precedence 
- [x] Documentation
- [x] Tests
- [x] Python bindings

Closes https://github.com/se-sic/VaRA/issues/702

s9latimm committed:
Implement constraint builder

A	include/vara/Feature/ConstraintBuilder.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/CMakeLists.txt
A	unittests/Feature/ConstraintBuilder.cpp

s9latimm commented:
@sarvech76

s9latimm committed:
Support parentheses

M	include/vara/Feature/ConstraintBuilder.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/ConstraintBuilder.cpp

s9latimm committed:
Add tests

M	unittests/Feature/ConstraintBuilder.cpp

s9latimm committed:
Add python bindings

A	bindings/python/tests/constraint/__init__.py
A	bindings/python/tests/constraint/test_constraint.py
M	bindings/python/vara-feature/CMakeLists.txt
A	bindings/python/vara-feature/pybind_Constraint.cpp
M	bindings/python/vara-feature/pybind_Init.cpp

s9latimm committed:
Add python tests

M	bindings/python/tests/constraint/test_constraint.py
M	bindings/python/vara-feature/pybind_Constraint.cpp
M	include/vara/Feature/ConstraintBuilder.h

s9latimm committed:
Fix tests

M	bindings/python/tests/constraint/test_constraint.py
M	unittests/Feature/ConstraintBuilder.cpp

vulder commented on include/vara/Feature/ConstraintBuilder.h:
```diff
@@ -0,0 +1,342 @@
+  ///
+  /// Example:
+  ///     left().feature("A").implies().feature("B").right()
+  ///      .implies().feature("C").implies().feature("D")
```
Why does this example have not left/right/ for the right side `(C => D)`

s9latimm replied:
Using the builder all operators are right-associative by default and parentheses inserted respectively, so we do not need to specify those here. It would be fine if one would do so, but not necessary.

vulder replied:
How would you, with this syntax then express this expression: `((A => B) => C => D)` and  `(((A => B) => C) => D)`

s9latimm replied:
You would need too explicitly parenthesize both expressions or simply use the `()` operator.

`feature("A").implies().feature("B")().implies().feature("C")().implies().feature("D")`

vulder reviewed (approved):
looks every nice overall, we can integrate this after fixing the few things discussed on friday.

nice work :)

s9latimm committed:
Refactor parenthesize

M	bindings/python/tests/constraint/test_constraint.py
M	bindings/python/vara-feature/pybind_Constraint.cpp
M	include/vara/Feature/ConstraintBuilder.h
M	unittests/Feature/ConstraintBuilder.cpp

s9latimm committed:
Add comment about parentheses

M	include/vara/Feature/ConstraintBuilder.h

s9latimm committed:
Merge remote-tracking branch 'origin/vara-dev' into f-ConstraintBuilder

M	bindings/python/tests/TEST_INPUTS/example_feature_model.xml
M	bindings/python/tests/TEST_INPUTS/simple_example_feature_model.xml
M	bindings/python/tests/TEST_INPUTS/simple_example_feature_model_merge.xml
M	bindings/python/vara-feature/CMakeLists.txt
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/FeatureModelTransaction.h
M	include/vara/Feature/FeatureModelWriter.h
M	include/vara/Feature/FeatureSourceRange.h
M	lib/Feature/FeatureModel.cpp
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	lib/Feature/XmlConstants.h
M	unittests/CMakeLists.txt
M	unittests/Feature/FeatureModel.cpp
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModelWriter.cpp
M	unittests/Feature/UnittestHelper.h
M	unittests/resources/CMakeLists.txt
M	unittests/resources/xml/test_children.xml
A	unittests/resources/xml/test_constraints.xml

s9latimm committed:
Add call bindings

M	bindings/python/tests/constraint/test_constraint.py
M	bindings/python/vara-feature/pybind_Constraint.cpp

s9latimm committed:
Add test

M	bindings/python/tests/constraint/test_constraint.py
M	unittests/Feature/ConstraintBuilder.cpp

s9latimm committed:
Add precedence

M	bindings/python/tests/constraint/test_constraint.py
M	include/vara/Feature/ConstraintBuilder.h
M	unittests/Feature/ConstraintBuilder.cpp

s9latimm committed:
Fix parentheses error

M	include/vara/Feature/ConstraintBuilder.h
